src/backend/replication/walsender.c

```diff
@@ -1288,10 +1296,7 @@ StartLogicalReplication(StartReplicationCmd *cmd)
 	 * are reported early.
 	 */
 	logical_decoding_ctx =
-		CreateDecodingContext(cmd->startpoint, cmd->options, false,
-							  XL_ROUTINE(.page_read = logical_read_xlog_page,
-										 .segment_open = WalSndSegmentOpen,
-										 .segment_close = wal_segment_close),
+		CreateDecodingContext(cmd->startpoint, cmd->options, false, &xlr,
```

Comment:
It's interesting that `CreateDecodingContext` accepts `&xlr` as a pointer, I wonder does it guarantees that it always copies the content and doesn't use the pointer after returning?

I found that it's copied at least in `XLogReaderAllocate`, but not sure if it's always the case.

Reply:
I think if this code is affected by this problem, the old code would also be affected. The XL_ROUTINE macro is defined as
```
#define XL_ROUTINE(...) &(XLogReaderRoutine){__VA_ARGS__}
```
So it creates an XLogReaderRoutine in the local stack frame anyway.